File: jar-shell/src/test/java/com/beancounter/shell/integ/TestPortfolioCommands.java
package com.beancounter.shell.integ;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.auth.client.AuthClientConfig;
import com.beancounter.auth.common.TokenService;
import com.beancounter.auth.common.TokenUtils;
import com.beancounter.client.services.PortfolioServiceClient;
import com.beancounter.client.services.PortfolioServiceClient.PortfolioGw;
import com.beancounter.client.services.RegistrationService;
import com.beancounter.common.contracts.PortfolioResponse;
import com.beancounter.common.contracts.PortfoliosRequest;
import com.beancounter.common.contracts.PortfoliosResponse;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.SystemUser;
import com.beancounter.common.utils.BcJson;
import com.beancounter.common.utils.KeyGenUtils;
import com.beancounter.common.utils.PortfolioUtils;
import com.beancounter.shell.cli.PortfolioCommands;
import com.beancounter.shell.config.ShellConfig;
import java.util.Collections;
import java.util.UUID;
import lombok.SneakyThrows;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.cloud.contract.stubrunner.spring.AutoConfigureStubRunner;
import org.springframework.cloud.contract.stubrunner.spring.StubRunnerProperties;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.test.context.ActiveProfiles;

@SpringBootTest(classes = {
    ShellConfig.class, AuthClientConfig.class
})
@AutoConfigureStubRunner(
    stubsMode = StubRunnerProperties.StubsMode.LOCAL,
    ids = "org.beancounter:svc-data:+:stubs:10999")
@ActiveProfiles("test")
public class TestPortfolioCommands {

  private static final TokenService tokenService = new TokenService();
  private static PortfolioCommands portfolioCommands;
  private static PortfolioGw portfolioGw;
  @MockBean
  private RegistrationService registrationService;
  @MockBean
  private JwtDecoder jwtDecoder;
  private final BcJson bcJson = new BcJson();

  @BeforeAll
  static void mockPortfolioGw() {
    portfolioGw = Mockito.mock(PortfolioGw.class);
    PortfolioServiceClient portfolioServiceClient =
        new PortfolioServiceClient(portfolioGw, tokenService);
    portfolioCommands = new PortfolioCommands(portfolioServiceClient);

  }

  @Test
  @SneakyThrows
  void getPortfolios() {
    getSystemUser();
    String result = portfolioCommands.get();
    assertThat(result).isNotBlank();
  }

  @Test
  @SneakyThrows
  void createPortfolio() {
    SystemUser owner = getSystemUser();

    PortfoliosResponse response = new PortfoliosResponse(
        Collections.singletonList(getPortfolio("ABC", owner)));

    Mockito.when(portfolioGw.getPortfolioByCode(
        Mockito.eq(tokenService.getBearerToken()),
        Mockito.isA(String.class)))
        .thenReturn(new PortfolioResponse(PortfolioUtils.getPortfolio("ABC")));

    Mockito.when(portfolioGw.addPortfolios(
        Mockito.eq(tokenService.getBearerToken()),
        Mockito.isA(PortfoliosRequest.class)))
        .thenReturn(response);

    String result = portfolioCommands
        .add("ABC", "ABC", "NZD", "USD");
    assertThat(result).isNotNull();
    Portfolio portfolio = bcJson.getObjectMapper().readValue(result, Portfolio.class);
    assertThat(portfolio).isEqualToIgnoringGivenFields(response.getData().iterator().next(),
        "owner");
  }

  @Test
  @SneakyThrows
  void is_AddPortfolioThatExists() {

    SystemUser owner = getSystemUser();
    Portfolio existing = getPortfolio("ZZZ", owner);
    PortfolioResponse portfolioResponse = new PortfolioResponse(existing);
    Mockito.when(portfolioGw.getPortfolioByCode(tokenService.getBearerToken(), existing.getCode()))
        .thenReturn(portfolioResponse); // Portfolio exists

    String result = portfolioCommands
        .add("ZZZ", "ABC", "NZD", "USD");
    assertThat(result).isNotNull();
    Portfolio portfolio = bcJson.getObjectMapper().readValue(result, Portfolio.class);
    assertThat(portfolio).isEqualToIgnoringGivenFields(portfolioResponse.getData(),
        "owner");
  }

  private SystemUser getSystemUser() {
    SystemUser owner = new SystemUser(KeyGenUtils.format(UUID.randomUUID()));
    Jwt jwt = TokenUtils.getUserToken(owner);
    Mockito.when(jwtDecoder.decode("token")).thenReturn(jwt);
    Mockito.when(registrationService.me()).thenReturn(owner);

    SecurityContextHolder.getContext().setAuthentication(
        new JwtAuthenticationToken(jwtDecoder.decode("token")));
    return owner;
  }

  private Portfolio getPortfolio(String code, SystemUser owner) {
    Portfolio toReturn = PortfolioUtils.getPortfolio(code);
    toReturn.setOwner(owner);
    return toReturn;
  }

}

File: jar-shell/src/test/kotlin/com/beancounter/shell/integ/TestPortfolioCommands.kt
package com.beancounter.shell.integ

import com.beancounter.auth.AutoConfigureMockAuth
import com.beancounter.auth.MockAuthConfig
import com.beancounter.auth.TokenService
import com.beancounter.auth.client.ClientPasswordConfig
import com.beancounter.client.services.PortfolioServiceClient
import com.beancounter.client.services.RegistrationService
import com.beancounter.common.contracts.PortfoliosRequest
import com.beancounter.common.contracts.PortfoliosResponse
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.model.Portfolio
import com.beancounter.common.model.SystemUser
import com.beancounter.common.utils.BcJson.Companion.objectMapper
import com.beancounter.common.utils.KeyGenUtils
import com.beancounter.common.utils.PortfolioUtils.Companion.getPortfolio
import com.beancounter.shell.Constants.Companion.NZD
import com.beancounter.shell.Constants.Companion.USD
import com.beancounter.shell.commands.PortfolioCommands
import com.beancounter.shell.config.ShellConfig
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.`when`
import org.mockito.kotlin.any
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.cloud.contract.stubrunner.spring.AutoConfigureStubRunner
import org.springframework.cloud.contract.stubrunner.spring.StubRunnerProperties
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean
import java.util.UUID

/**
 * Verify portfolio commands.
 */
@SpringBootTest(classes = [ShellConfig::class, ClientPasswordConfig::class])
@AutoConfigureStubRunner(
    stubsMode = StubRunnerProperties.StubsMode.LOCAL,
    ids = ["org.beancounter:svc-data:0.1.1:stubs:10991"]
)
@ActiveProfiles("jar-shell-shared", "contract-base")
@AutoConfigureMockAuth
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
class TestPortfolioCommands {
    @MockitoBean
    private lateinit var registrationService: RegistrationService

    @MockitoBean
    private lateinit var jwtDecoder: JwtDecoder

    @MockitoBean
    private lateinit var portfolioServiceClient: PortfolioServiceClient

    @Autowired
    private lateinit var mockAuthConfig: MockAuthConfig

    @Autowired
    private lateinit var tokenService: TokenService

    private lateinit var portfolioCommands: PortfolioCommands

    @BeforeEach
    fun configure() {
        portfolioCommands = PortfolioCommands(portfolioServiceClient)
    }

    private val pfCode = "ABC"

    @BeforeEach
    fun autoLogin() {
        mockAuthConfig.login()
    }

    @Test
    fun createPortfolio() {
        val owner = systemUser
        val response =
            PortfoliosResponse(
                listOf(
                    getPortfolio(
                        pfCode,
                        owner
                    )
                )
            )
        // Throw exception so creation path is taken
        `when`(portfolioServiceClient.getPortfolioByCode(anyString()))
            .thenThrow(BusinessException("Not found"))

        `when`(portfolioServiceClient.add(any<PortfoliosRequest>()))
            .thenReturn(response)

        val result =
            portfolioCommands.add(
                pfCode,
                pfCode,
                NZD.code,
                USD.code
            )
        assertThat(result).isNotNull
        val portfolio =
            objectMapper.readValue(
                result,
                Portfolio::class.java
            )
        assertThat(portfolio)
            .usingRecursiveComparison()
            .ignoringFields("owner")
            .isEqualTo(response.data.iterator().next())
    }

    @Test
    fun is_AddPortfolioThatExists() {
        val owner = systemUser
        val code = "ZZZ"
        val existing =
            getPortfolio(
                code,
                owner
            )
        `when`(portfolioServiceClient.getPortfolioByCode(existing.code))
            .thenReturn(existing) // Portfolio exists
        val result =
            portfolioCommands.add(
                code,
                pfCode,
                NZD.code,
                USD.code
            )
        assertThat(result).isNotNull
        val portfolio =
            objectMapper.readValue(
                result,
                Portfolio::class.java
            )
        assertThat(portfolio).usingRecursiveComparison().isEqualTo(existing)
    }

    private val systemUser: SystemUser
        get() {
            return SystemUser(KeyGenUtils().format(UUID.randomUUID()))
        }

    private fun getPortfolio(
        code: String,
        owner: SystemUser
    ): Portfolio {
        val toReturn = getPortfolio(code)
        toReturn.owner = owner
        return toReturn
    }
}